### PR TITLE
Android: Make build scripts more reliable

### PR DESCRIPTION
**Pull request by alexander255**
Currently the Android app build scripts will not detect errors, causing spurious follow-up issues (such as configure failing with an *libcurl not found* error after libcurl failed to build or be correctly found, or `strip` dump a sh*load of spurious errors because the `cd` to the binary directory failed and it tries to parse the top-level source and text files instead, etc.).
Also it's very hard for somebody calling `./build_all.sh` from a script to figure out whether the build *actually* succeeded or not (probing the app binary directory works, but is wokey)…

### CODE REVIEW

**Comment by adamradocz** [comment]
This PR maybe solve the #1989 issue?

**Comment by alexander255** [comment]
Unlikely, the BOINC Android app will compile and install just fine when the built binaries are broken/missing: Android platform simply treats the binaries as “static files”/assets and basically ignores them on install; only resources and Java bytecode (“classes.dex”) get special treatment iirc. The app *will* **hang** on startup in this case through, while filling `logcat` with “could not start client”-style messages.

No, if I had to guess what *his* issue is, I'd say he probably simply did a release build of the BOINC client app and forgot to (ie: didn't know that he has to) sign it. Only debug builds are signed by default using the standard debug keys, release builds are unsigned by default and need to be signed (using any key) before installation.

(For installation errors one usually receives much more useful error messages when attempting to install the app using `adb install <.apk>` instead of using the PackageManager GUI.)

**Comment by TheAspens** [comment]
@alexander255 - I'm working on tracking down someone to review this code.  Both people I've asked are out this week on vacation but said they would try to look at it next week.  Sorry for the delay in getting it reviewed.

**Comment by AenBleidd** [comment]
I think about adding Android build to CI. Will try to do this during this
week.

Best regards,
Vitalii Koshura

2018-02-05 16:33 GMT+02:00 Kevin Reed <notifications@github.com>:

**Comment by TheAspens** [comment]
@AenBleidd - that would be great.  Thanks!

File: android/build_androidtc_arm.sh [hunk 1]
```diff
@@ -10,6 +10,6 @@ export NDK_ROOT="${NDK_ROOT:-$HOME/NVPACK/android-ndk-r10e}"
 export ANDROID_TC="${ANDROID_TC:-$HOME/android-tc}"
 export ANDROID_TC_ARM="${ANDROID_TC_ARM:-$ANDROID_TC/arm}"
 
-if [ ! -d $ANDROID_TC_ARM/arm-linux-androideabi ]; then
-    $NDK_ROOT/build/tools/make-standalone-toolchain.sh --platform=android-9 --arch=arm --install-dir=$ANDROID_TC_ARM
+if [ ! -d "$ANDROID_TC_ARM/arm-linux-androideabi" ]; then
```
Comment (brevilo) [inline android/build_androidtc_arm.sh]: One could argue that this is redundant but, ok, it's future-proof should line 11 change or be removed.

File: android/build_libraries_arm.sh [hunk 2]
```diff
@@ -28,21 +28,21 @@ export CFLAGS="--sysroot=$TCSYSROOT -DANDROID -DDECLARE_TIMEZONE -Wall -I$TCINCL
...
 echo "==================building Libraries from $BOINC=========================="
-cd $BOINC
+cd "$BOINC"
 if [ -n "$MAKECLEAN" ]; then
 make clean
 fi
 if [ -n "$CONFIGURE" ]; then
 ./_autosetup
-./configure --host=arm-linux --with-boinc-platform="arm-android-linux-gnu" --prefix=$TCINCLUDES --libdir="$TCINCLUDES/lib" --with-ssl=$TCINCLUDES --disable-server --disable-manager --disable-client --disable-shared --enable-static --enable-boinczip
+./configure --host=arm-linux --with-boinc-platform="arm-android-linux-gnu" --prefix="$TCINCLUDES" --libdir="$TCINCLUDES/lib" --with-ssl=$TCINCLUDES --disable-server --disable-manager --disable-client --disable-shared --enable-static --enable-boinczip
```
Comment (brevilo) [inline android/build_libraries_arm.sh]: You missed `--with-ssl=$TCINCLUDES`

File: android/build_boinc_arm.sh [hunk 3]
```diff
@@ -27,20 +27,20 @@ export CFLAGS="--sysroot=$TCSYSROOT -DANDROID -DDECLARE_TIMEZONE -Wall -I$TCINCL
 export CXXFLAGS="--sysroot=$TCSYSROOT -DANDROID -Wall -I$TCINCLUDES/include -funroll-loops -fexceptions -O3 -fomit-frame-pointer -fPIE -march=armv7-a"
```
Comment (brevilo) [inline android/build_boinc_arm.sh]: What's the reasoning behind not improving the contents of `CXXFLAGS`, `LDFLAGS` and `GDB_CFLAGS` in the same way (in all scripts)? I'm not sure the outer quotation marks are effective for the contained inner variables like in `-I$TCINCLUDES/include` as well. Are they?

File: android/build_libraries_x86.sh [hunk 4]
```diff
@@ -14,7 +14,7 @@ MAKECLEAN="yes"
 export BOINC=".." #BOINC source code
 
 export ANDROID_TC="${ANDROID_TC:-$HOME/android-tc}"
-export ANDROIDTC="${ANDROID_TC_X86:-$ANDROID_TC/x86"
+export ANDROIDTC="${ANDROID_TC_X86:-$ANDROID_TC/x86}"
 export TCBINARIES="$ANDROIDTC/bin"
```
Comment (brevilo) [inline android/build_libraries_x86.sh]: Why not treat the following variables like `ANDROIDTC` above? Yes, that's already fixed but you keep depending on that. Ideally you want to protect any variable, without further dependencies.

File: android/build_libraries_arm.sh (hunk 2 above)
Comment (alexander255) [reply android/build_libraries_arm.sh]: “Made with love by hand by …” – Guess I missed that one. Fixed and checked them all again.

File: android/build_boinc_arm.sh (hunk 3 above)
Comment (alexander255) [reply android/build_boinc_arm.sh]: Interesting question actually… I so routinely ignore this issue I didn't have an imediate answer for this.

Turns out the answer is twofold:

  * As long as you use `autotools` there is absolutely nothing you can do here – any attempt at adding quotation or similar will make things worse! And [`cmake` is in no way better](https://gitlab.kitware.com/cmake/cmake/blob/master/Modules/CMakeCInformation.cmake#L103) if I understand this code correctly.
  * I also checked `meson` and [it uses Python's `shlex.split`](https://github.com/mesonbuild/meson/blob/ae0e469b0692862dae10ceb02799feb3f117ce26/mesonbuild/environment.py#L924), which means it would support quoting the values correctly.

TL;DR: Yes this sucks and there is no way to fix it currently.

File: android/build_libraries_x86.sh (hunk 4 above)
Comment (alexander255) [reply android/build_libraries_x86.sh]: Sorry, I don't understand what you are saying. Protect what variable, with which dependencies?

**Comment by alexander255** [comment]
Hopefully all fixable issues (including `#!/bin/sh -e` vs `#!/bin/sh\nset -e`) should be fixed now.

**Comment by truboxl** [comment]
Is it possible to correct the shebangs in all the scripts once and for all (missing !) ?
Probably a very small matter...

**Comment by brevilo** [comment]
Yet an important one that was missed up to now! Nice catch!

**Comment by alexander255** [comment]
@truboxl: Could you point out an example for a script that doesn't have a shebang?

**Comment by brevilo** [comment]
As he wrote, they all lack the "!".

**Comment by alexander255** [comment]
@brevilo: I looked at the scripts yesterday and I literally didn't *see* that. – `sed` fixed it just now.

**Comment by brevilo** [comment]
Yeah, me too, obviously 😎

**Comment by ChristianBeer** [comment]
I commented in #2409 about this but I want to raise the question here also: Why do you want to use `set -e` versus `if [ $? -ne 0 ]; then exit 1; fi` as the latter is done in all the CI scripts for Mac and Linux already. Why not be consistent and also use it for android? At least in scripts that use `if` construct the `set -e` should be removed.

**Comment by AenBleidd** [comment]
@ChristianBeer, @alexander255, 
When I tested these scripts with `sh -e` on travis ci it always show no errors even when script failed.
I didn't check this on local environment but `if [ $? -ne 0 ]; then exit 1; fi` works well

**Comment by alexander255** [comment]
@AenBleidd: Attempting to build BOINC using `./build_all.sh` without setting `"${OPENSSL_SRC}` properly copies all the ToolChain stuff and then “cleanly” fails with `can't cd to ~/src/openssl-1.0.2g` on the first error, setting the exit code (`$?`) to `2` on my system (`/bin/sh` is `dash`).

… and the reason is of course because I forgot to add the changed files before commiting and pushing as @truboxl correctly pointed out just now. :weary:

**Comment by AenBleidd** [comment]
@alexander255, maybe we can cooperate here somehow?
I did the similar but more extended work on these scripts so maybe we can collect all fixes and make 1 clean pull request with all your changes as one squashed commit and all my changes on top of your changes as one squashed commit (2 commits in total)?
I think this can speed-up development and approval process.
Also this can save my time and your time.
What do you think?

**Comment by alexander255** [comment]
Sure, I don't care. I just submitted these as I had encountered some issues when building from source and wanted to upstream it. Basically you're asking me to squash my commits so that you can reuse them as part of your PR, is that correct? If so I'd prefer to keep the first commit seperate as it addresses a specific problem that is described in detail in the commit message, while the others just fix “generic code problems”. Is that OK for you?

**Comment by AenBleidd** [comment]
I'm ok with any possible ways. I just want to help you not to make the work I did already.
In any case I'll squash my commits on top of your code and resolve all conflicts.
As from my POV, it is better to review both your and my changes in scope as far as they contain changes to the same scripts.
If other guys are ok with this - I'd do it this way.
If no - I'm ok to keep them as a separate PRs.

**Comment by alexander255** [comment]
@AenBleidd: Done! Please go ahead! :slightly_smiling_face:

**Comment by brevilo** [comment]
@alexander255 there's still one review question pending regarding `android/build_libraries_x86.sh`.

File: android/build_build_app_x86.sh [hunk 5]
```diff
@@ -25,7 +26,7 @@ export GDB_CFLAGS="--sysroot=$TCSYSROOT -Wall -g -I$TCINCLUDES/include"
 make clean
 
 if [ -e "./confgiure" ]; then
```
Comment (JuhaSointusalo) [inline android/build_build_app_x86.sh]: Could you fix this typo too?

**Comment by JuhaSointusalo** [comment]
`set -e` has some pitfalls. See [this Server Fault answer](https://serverfault.com/a/847016) for examples as well as the linked pages.

I'm not saying you can't use `set -e` and these scripts are simple and probably don't have any issues but I think anyone working on these scripts need to be aware of the pitfalls.

**Comment by alexander255** [comment]
@JuhaSointusalo: Fixed!

**Comment by AenBleidd** [comment]
@brevilo, @JuhaSointusalo, are you ok with the latest changes?

**Comment by JuhaSointusalo** [comment]
@AenBleidd The changes look ok to me. I don't have Android build environment set up to test this so I'll defer to @brevilo.

**Comment by AenBleidd** [comment]
@JuhaSointusalo, you can take a look at my WIP PR 2437 that based on the latest changes from this PR but with my CI scripts and some additional changes. I plan to squash that PR when this one will be merged into master

**Comment by alexander255** [comment]
I hate to be begging like this, but could somebody representative of the project give me a response the my second question on #2393. My [inclusion request](https://gitlab.com/fdroid/fdroiddata/merge_requests/2861) into the F-Droid repo has been open for over 2 months at this point and is currently dead-locked on that question.

**Comment by brevilo** [comment]
@AenBleidd I don't even know what you mean by "latest" since none of the updates to this PR were done via incremental commits, unfortunately. As to me previous review, my [last comment](https://github.com/BOINC/boinc/pull/2330#issuecomment-376079377) still holds.

**Comment by brevilo** [comment]
@alexander255 What *exactly* do you mean by "my second question on #2393". The second question or the second bullet point - presuming we're talking about the opening comment, as your second comment also contains a question. Thanks.

**Comment by AenBleidd** [comment]
> @AenBleidd I don't even know what you mean by "latest" since none of the updates to this PR were done via incremental commits, unfortunately. As to me previous review, my last comment still holds.

@brevilo, I've just created new PR with new code and test on it. I also don't like this way of making commits but I used to work with different cases so it's not a problem for me to do all job one more time

**Comment by alexander255** [comment]
@brevilo: Mostly second bullet, but I already got a detailed answer to both questions by CristianBeer. It's not really related to this issue either way.

**Comment by AenBleidd** [comment]
@brevilo, any updates?

**Comment by brevilo** [comment]
Sorry, GDPR all over I guess. I'm closing my review as the remaining questions aren't that important and the discussion just lost coherence for me...

**Comment by AenBleidd** [comment]
@truboxl, are you ok with the changes now?

File: android/build_libraries_x86.sh (hunk 4 above)
Comment (alexander255) [reply android/build_libraries_x86.sh]: While I always add curly braces (`{}`) around all variables in shell code *I* write, doing so is only required when doing any kind of variable substitution (ie: `${ANDROID_TC_X86:-Default}`). In practice you'll find lots and lots of shell code that only uses the braces when required and this won't cause any problems if you always consider wether the character *after* the variable name *may be part of a shell variable name* or not when writing it.

I tend to ofter *not* consider the above criterion in practice so I add the braces just to be safe, but this really only matters when you write the code; once it's properly written adding the braces everywhere is really not worth the (wasted) time, while also likely to cause hard to spot errors if the code is not reviewed very carefully (wasting everybody's time).

**Comment by TheAspens** [comment]
Merging based on approval of reviewers and contributors
